File: ParalellAndVectorizedMatrices/src/main/java/org/example/ExecutorMatrixMultiplication.java
package org.example;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ExecutorMatrixMultiplication implements ParallelMatrixMultiplication{

    @Override
    public double[][] multiply(double[][] A, double[][] B, int threads) throws InterruptedException {
        int rowsA = A.length;
        int colsA = A[0].length;
        int colsB = B[0].length;
        double[][] result = new double[rowsA][colsB];

        ExecutorService executor = Executors.newFixedThreadPool(threads);

        for (int i = 0; i < rowsA; i++) {
            int row = i;
            executor.execute(() -> {
                for (int j = 0; j < colsB; j++) {
                    double sum = 0;
                    for (int k = 0; k < colsA; k++) {
                        sum += A[row][k] * B[k][j];
                    }
                    result[row][j] = sum;
                }
            });
        }

        executor.shutdown();
        while (!executor.isTerminated()) {
            Thread.sleep(10);
        }

        return result;
    }
}

File: ParallelAndVectorized/Algorithms/ExecutorMatrixMultiplication.py
import numpy as np
from concurrent.futures import ThreadPoolExecutor

class ExecutorMatrixMultiplication:
    def multiply(self, A, B, threads):
        A = np.array(A, dtype=np.float32)
        B = np.array(B, dtype=np.float32)

        rowsA, colsA = A.shape
        colsB = B.shape[1]

        result = np.zeros((rowsA, colsB), dtype=np.float32)

        def process_row(row):
            for j in range(colsB):
                result[row, j] = np.dot(A[row, :], B[:, j])

        with ThreadPoolExecutor(max_workers=threads) as executor:
            executor.map(process_row, range(rowsA))

        return result
